ipsec: Fix incorrect trace in from-network

YutaroHayakawa opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

In from-network, there is a wrong send_trace_notify which sets
TRACE_REASON_ENCRYPTED for decrypted packets.

```release-note
Fix incorrect packet trace for encrypted packets received from the network
```

pchaigno commented on bpf/bpf_network.c:
```diff
@@ -19,8 +19,7 @@ int from_network(struct __ctx_buff *ctx)
 		send_trace_notify(ctx, TRACE_FROM_NETWORK, get_identity(ctx), 0, 0,
-				  ctx->ingress_ifindex,
-				  TRACE_REASON_ENCRYPTED, TRACE_PAYLOAD_LEN);
+				  ctx->ingress_ifindex, 0, TRACE_PAYLOAD_LEN);
```
I think we have [the same issue at `bpf_overlay`](https://github.com/cilium/cilium/blob/v1.11.1/bpf/bpf_overlay.c#L306-L308).

YutaroHayakawa replied:
Ok, let me fix that together.

pchaigno commented on bpf/bpf_network.c:
```diff
@@ -19,8 +19,7 @@ int from_network(struct __ctx_buff *ctx)
 		send_trace_notify(ctx, TRACE_FROM_NETWORK, get_identity(ctx), 0, 0,
-				  ctx->ingress_ifindex,
-				  TRACE_REASON_ENCRYPTED, TRACE_PAYLOAD_LEN);
+				  ctx->ingress_ifindex, 0, TRACE_PAYLOAD_LEN);
```
Do we need to instead emit the `TRACE_REASON_ENCRYPTED` trace below, for when packets come in without a mark? We may need to parse the packet to check that it's actually an ESP packet as we may also receive unencrypted packets.

YutaroHayakawa replied:
Yeah, I was wondering if it is worth it to make sure the packet is encrypted or not in here just for the tracing, but we anyways do that with `do_decrypt` here. I'll try to refactor do_decrypt to return protocol number then.

YutaroHayakawa replied:
Fixed in 9ca7d1170dc5eb96a06fd4f8aad0e402cfe9d52b. Finally, I didn't change `do_decrypt`.

pchaigno reviewed (commented):
> Fix incorrect trace in from-network

`from-network` won't mean anything to users (though they could guess :) ). What about `Fix incorrect packet trace for encrypted packets received from the network`?

pchaigno commented on bpf/bpf_network.c:
```diff
@@ -34,11 +26,49 @@ int from_network(struct __ctx_buff *ctx)
 		return CTX_ACT_OK;
 
-	return do_decrypt(ctx, proto);
+	act = do_decrypt(ctx, proto);
```
Nit: We usually call that variable `ret`.

pchaigno reviewed (commented):
I think the third commit is enough (changes made in the first commit are overwritten by the third anyway). Also, don't we need the same fix as in commit 3 for bpf_overlay? Changes in bpf_network look good to me :+1:

YutaroHayakawa committed:
ipsec: Trace ESP packets coming from the network as encrypted

When IPSec is enabled, it is helpful to trace ESP packets as encrypted.
Fix the current program to do that. In non-IPSec mode, it won't happen
because we are not interested in the encrypted packets. Please see the
comment as well for more details.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
ipsec: Trace ESP packets coming from the overlay as encrypted

When IPSec is enabled, it is helpful to trace ESP packets as encrypted.
Fix the current program to do that. In non-IPSec mode, it won't happen
because we are not interested in the encrypted packets. Please see the
comment as well for more details.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

pchaigno commented on bpf/bpf_overlay.c:
```diff
@@ -310,6 +310,41 @@ int tail_handle_ipv4(struct __ctx_buff *ctx)
+	struct ipv6hdr *ip6;
+#endif
+#ifdef ENABLE_IPV4
+	struct iphdr *ip4;
```
We usually simply add `__maybe_unused` instead of macros for the variables. Otherwise it quickly becomes very heavy in macros :smile_cat:

pchaigno commented on bpf/bpf_overlay.c:
```diff
@@ -310,6 +310,41 @@ int tail_handle_ipv4(struct __ctx_buff *ctx)
+		return false;
+	}
+
+	return (protocol == IPPROTO_ESP);
```
Nit: Parenthesis are not needed here.

pchaigno reviewed (approved):
Only minor comments this time.

If you only push fixes for those without rebasing, I don't think we'll need to rerun the end-to-end tests (assuming they pass now).
